controlplane/kubeadm/internal/cluster.go

```diff
@@ -70,17 +87,25 @@ func OwnedControlPlaneMachines(controlPlaneName string) func(machine *clusterv1.
 // that have a deletion timestamp.
 func HasDeletionTimestamp() func(machine *clusterv1.Machine) bool {
 	return func(machine *clusterv1.Machine) bool {
-		return machine.GetDeletionTimestamp() != nil
+		return !WithoutDeletionTimestamp()(machine)
 	}
 }
 
-// HasOutdatedConfiguration returns a MachineFilter function to find all machines
-// that do not match a given KubeadmControlPlane configuration hash.
-func HasOutdatedConfiguration(configHash string) func(machine *clusterv1.Machine) bool {
+// WithoutDeletionTimestamp returns a MachineFilter function to find all machines
+// that do not have a deletion timestamp.
+func WithoutDeletionTimestamp() func(machine *clusterv1.Machine) bool {
```

Comment:
Consider either renaming to `DeletionTimestampIsZero` or implement a `Not()` function that takes a `func(machine *clusterv1.Machine) bool` and returns a `func(machine *clusterv1.Machine) bool` that returns the opposite of the function passed in. 

Then the implementation found in `WithoutDeletionTimestamp` can move into `HasDeletionTimestamp` and this could call `FilterMachines(machines, Not(HasDeletionTimestamp()))`

This would also provide `Not(MatchesConfigurationHash(cfg))` without needing to define `HasOutdatedConfiguration`.